src/parallel_prover.rs

```diff
@@ -336,48 +334,151 @@ where
     if self.i_end + 1 != right.i_start {
-        return Err(NovaError::InvalidNodeMerge);
+      return Err(NovaError::InvalidNodeMerge);
     }
 
     // First we fold the secondary instances of both the left and right children
     let (nifs_left_secondary, (left_U_secondary, left_W_secondary)) = NIFS::prove(
-        &pp.ck_secondary,
-        &pp.ro_consts_secondary,
-        &pp.r1cs_shape_secondary,
-        &self.r_U_secondary,
-        &self.r_W_secondary,
-        &self.l_u_secondary,
-        &self.l_w_secondary,
+      &pp.ck_secondary,
+      &pp.ro_consts_secondary,
+      &pp.r1cs_shape_secondary,
+      &self.r_U_secondary,
+      &self.r_W_secondary,
+      &self.l_u_secondary,
+      &self.l_w_secondary,
     )?;
     let (nifs_right_secondary, (right_U_secondary, right_W_secondary)) = NIFS::prove(
-        &pp.ck_secondary,
-        &pp.ro_consts_secondary,
-        &pp.r1cs_shape_secondary,
-        &right.r_U_secondary,
-        &right.r_W_secondary,
-        &right.l_u_secondary,
-        &right.l_w_secondary,
+      &pp.ck_secondary,
+      &pp.ro_consts_secondary,
+      &pp.r1cs_shape_secondary,
+      &right.r_U_secondary,
+      &right.r_W_secondary,
+      &right.l_u_secondary,
+      &right.l_w_secondary,
     )?;
     let (nifs_secondary, (U_secondary, W_secondary)) = NIFS::prove(
-        &pp.ck_secondary,
-        &pp.ro_consts_secondary,
-        &pp.r1cs_shape_secondary,
-        &left_U_secondary,
-        &left_W_secondary,
-        &right_U_secondary,
-        &right_W_secondary,
+      &pp.ck_secondary,
+      &pp.ro_consts_secondary,
+      &pp.r1cs_shape_secondary,
+      &left_U_secondary,
+      &left_W_secondary,
+      &right_U_secondary,
+      &right_W_secondary,
     )?;
 
-
     return Err(NovaError::InvalidStepOutputLength);
-  } 
-}
\ No newline at end of file
+  }
+}
+
+/// Structure for parallelization
+#[derive(Debug, Clone)]
+pub struct ParallelSNARK<G1, G2, C1, C2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+  C2: StepCircuit<G2::Scalar>,
+{
+  nodes: Vec<Vec<NovaTreeNode<G1, G2, C1, C2>>>,
+}
+
+/// Implementation for parallelization SNARK
+impl<G1, G2, C1, C2> ParallelSNARK<G1, G2, C1, C2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+  C2: StepCircuit<G2::Scalar>,
+{
+  /// Create a new instance of parallel SNARK
+  pub fn new(
+    pp: &PublicParams<G1, G2, C1, C2>,
+    steps: usize,
+    z0_primary: Vec<G1::Scalar>,
+    z0_secondary: Vec<G2::Scalar>,
+    c_primary: C1,
+    c_secondary: C2,
+  ) -> Self {
+    let mut zi = Vec::<(usize, Vec<G1::Scalar>, Vec<G2::Scalar>)>::new();
+    zi.push((0, z0_primary.clone(), z0_secondary.clone()));
+    for i in 1..steps {
+      let (index, prev_primary, prev_secondary) = &zi[i - 1];
```

Comment:
I think instead of pushing stored indexes to the vec it may be cleaner to use the .enumerate method in iter